README.md
[![Build (master)](https://github.com/filipdutescu/renamer.nvim/actions/workflows/ci_master.yaml/badge.svg)](https://github.com/filipdutescu/renamer.nvim/actions/workflows/ci_master.yaml)
[![Build](https://github.com/filipdutescu/renamer.nvim/actions/workflows/ci.yaml/badge.svg)](https://github.com/filipdutescu/renamer.nvim/actions/workflows/ci.yaml)
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/filipdutescu/renamer.nvim)](https://github.com/filipdutescu/renamer.nvim/releases)

# renamer.nvim

![Demo](https://j.gifs.com/qQAYmD.gif)

`renamer.nvim` is a Visual-Studio-Code-like renaming UI for Neovim, writen in
Lua. It is considerably customizable and uses the [Neovim >= 0.5.0](https://github.com/neovim/neovim/releases/tag/v0.5.0)
LSP feature as its backend.

## Table of contents

- [Features](#features)
- [Getting started](#getting-started)
- [Usage](#usage)
- [Customization](#customization)
- [Default mappings](#default-mappings)
- [Media](#media)
- [Contributing](#contributing)
- [License](#license)

## Features

- **Lightweight:** the overhead of this plugin is insignificant and it makes use
  of existent features or plugins you most likely already have installed (only
  [plenary.nvim](https://github.com/nvim-lua/plenary.nvim) required).
- **Responsive UI:** takes into account the cursor position and where the popup
  will be place relative to the current window to adjust the its placement.
- **[Neovim >= 0.5.0](https://github.com/neovim/neovim/releases/tag/v0.5.0) LSP**:
  uses the Neovim LSP to rename across scopes and project.
- **Popup customization**: provides several ways to integrate the popup with
  your specific setup, from border characters and title to its colours.
- **References highlighting:** of the word to be renamed, using Neovim's LSP (see
  [References highlighting](#references-highlighting))
- **Keymaps:** predefined keymaps to help you navigate and manipulate the popup
- **LSP handler:** if you do not wish to use the UI offered by the plugin, you can
still use the custom handler it has underneath to benefit from quickfix list
population and addition of your own custom response handler or cursor setting at
the end of the new word.

## Getting started

In order to start using `renamer.nvim` you will need to follow the following
sections. If you already meet certain requirements (such as already having
[plenary.nvim](https://github.com/nvim-lua/plenary.nvim)), then you only need to
install `renamer.nvim`.

### Prerequisites

To use `renamer.nvim`, since it makes use of
[Neovim](https://github.com/neovim/neovim)'s built-in LSP, you will need to have
installed [Neovim v0.5.0](https://github.com/neovim/neovim/releases/tag/v0.5.0)
or [newer](https://github.com/neovim/neovim/releases/latest).

### Installation

Using [vim-plug](https://github.com/junegunn/vim-plug):

```viml
Plug 'nvim-lua/plenary.nvim'
Plug 'filipdutescu/renamer.nvim', { 'branch': 'master' }
```

Using [dein](https://github.com/Shougo/dein.vim)

```viml
call dein#add('nvim-lua/plenary.nvim')
call dein#add('filipdutescu/renamer.nvim', { 'rev': 'master' })
```
Using [packer.nvim](https://github.com/wbthomason/packer.nvim)

```lua
use {
  'filipdutescu/renamer.nvim',
  branch = 'master',
  requires = { {'nvim-lua/plenary.nvim'} }
}
```

### Verify your installation

After those steps, you should be able to run `:checkhealth renamer` in order to
see if anything is missing from your setup.

After following the steps above, either continue reading below or run `:help
renamer` to get an understanding of the next steps required to use `renamer.nvim`
and how to configure it.

## Usage

To rename the current word using `renamer.nvim`, you need to call the `rename`
method (`require('renamer').rename()`).

The recommended way of doing it is by setting up keybindings to call the function:

VimScript:

```viml
inoremap <silent> <F2> <cmd>lua require('renamer').rename()<cr>
nnoremap <silent> <leader>rn <cmd>lua require('renamer').rename()<cr>
vnoremap <silent> <leader>rn <cmd>lua require('renamer').rename()<cr>
```

Lua:

```lua
vim.api.nvim_set_keymap('i', '<F2>', '<cmd>lua require("renamer").rename()<cr>', { noremap = true, silent = true })
vim.api.nvim_set_keymap('n', '<leader>rn', '<cmd>lua require("renamer").rename()<cr>', { noremap = true, silent = true })
vim.api.nvim_set_keymap('v', '<leader>rn', '<cmd>lua require("renamer").rename()<cr>', { noremap = true, silent = true })
```

## Customization

`renamer.nvim` offers different customization options, in order to change its
appearance and behaviour. Below you will find the defaults and the structure
of the settings table.

### Renamer setup structure

```lua
local mappings_utils = require('renamer.mappings.utils')
require('renamer').setup {
    -- The popup title, shown if `border` is true
    title = 'Rename',
    -- The padding around the popup content
    padding = {
        top = 0,
        left = 0,
        bottom = 0,
        right = 0,
    },
    -- The minimum width of the popup
    min_width = 15,
    -- The maximum width of the popup
    max_width = 45,
    -- Whether or not to shown a border around the popup
    border = true,
    -- The characters which make up the border
    border_chars = { '─', '│', '─', '│', '╭', '╮', '╯', '╰' },
    -- Whether or not to highlight the current word references through LSP
    show_refs = true,
    -- Whether or not to add resulting changes to the quickfix list
    with_qf_list = true,
    -- Whether or not to enter the new name through the UI or Neovim's `input`
    -- prompt
    with_popup = true,
    -- The keymaps available while in the `renamer` buffer. The example below
    -- overrides the default values, but you can add others as well.
    mappings = {
        ['<c-i>'] = mappings_utils.set_cursor_to_start,
        ['<c-a>'] = mappings_utils.set_cursor_to_end,
        ['<c-e>'] = mappings_utils.set_cursor_to_word_end,
        ['<c-b>'] = mappings_utils.set_cursor_to_word_start,
        ['<c-c>'] = mappings_utils.clear_line,
        ['<c-u>'] = mappings_utils.undo,
        ['<c-r>'] = mappings_utils.redo,
    },
    -- Custom handler to be run after successfully renaming the word. Receives
    -- the LSP 'textDocument/rename' raw response as its parameter.
    handler = nil,
}
```

Colours can also be changed, in order to better fit your theme or taste. The
following highlight groups can also be modified to theme the popup colours:

```viml
hi default link RenamerNormal Normal
hi default link RenamerBorder RenamerNormal
hi default link RenamerTitle Identifier
```

### References highlighting

`renamer.nvim` also **highlights the current text** wherever it is used in the
current document, using `vim.lsp.buf.document_highlight()`. In order for it to
be visible, the `LspReferenceText`, `LspReferenceRead` and `LspReferenceWrite`
highlight groups need to be defined (see the
[`vim.lsp.buf.document_highlight()`](https://neovim.io/doc/user/lsp.html#vim.lsp.buf.document_highlight())
documentation for details).

### Renamer rename structure

```lua
require('renamer').rename {
    -- Defines whether or not the popup should contain the initial word or be
    -- empty
    empty = false,
}
```

## Default mappings

Mappings are fully customizable. The default mappings are intended to be familiar
to users and mimic the default behaviour of their normal mode equivalents.

| Mappings | Action                       |
|----------|------------------------------|
| `<c-i>`  | Go to the start of the line. |
| `<c-a>`  | Go to the end of the line.   |
| `<c-e>`  | Go to the end of the word.   |
| `<c-b>`  | Go to the start of the word. |
| `<c-c>`  | Clear the buffer line.       |
| `<c-u>`  | Undo changes.                |
| `<c-r>`  | Redo undone changes.         |

To change existing mappings or add new ones, the value needed to be passed in the
mappings field is a Lua function. Here is an example of how to change `<c-c>`'s
behaviour from clearing the line to closing the popup:

```lua
-- other setup code...

['<c-c>'] = function()
    vim.api.nvim_input('<esc>')
end

-- other setup code...
```

## Media

Here are some examples of various parts of `renamer.nvim` that you can
customize, to hopefully give you a better idea of how to integrate the plugin
and maybe convince you to start using it.

- **Demo:**

![Demo](https://j.gifs.com/qQAYmD.gif)

- **Border customization:**

![border-chars](media/renamer_border_chars.png "Border customization")

or no border:

![no-border](media/renamer_no_border.png "No border")

- **Highlight references:**

![highlight-references](media/renamer_highlight_references.png "Highlight references")

- **No title:**

![no-title](media/renamer_no_title.png "No title")

- **No popup:**

![no-popup](media/renamer_no_popup.png "No popup")

Thanks [@sdushantha](https://github.com/sdushantha) for suggesting this (and
providing the screenshot)!

## Contributing

All contributions are welcome! Just open a pull request or an issue. Please read
[CONTRIBUTING.md](CONTRIBUTING.md).

## License

This project is licensed under the
[Apache 2.0 License](https://www.apache.org/licenses/LICENSE-2.0) - see the
[LICENSE](LICENSE) file for details



CHANGELOG.md
# [5.0.0](https://github.com/filipdutescu/renamer.nvim/compare/v4.0.1...v5.0.0) (2022-05-19)


### Bug Fixes

* **submit:** trim new word before using (GH-123) ([7a2c0a1](https://github.com/filipdutescu/renamer.nvim/commit/7a2c0a1fd47d3dc94cc77db27b94a586ccc007a1)), closes [#123](https://github.com/filipdutescu/renamer.nvim/issues/123) [#119](https://github.com/filipdutescu/renamer.nvim/issues/119)


### Features

* **handler:** migrate to handler (GH-122) ([6c99bd9](https://github.com/filipdutescu/renamer.nvim/commit/6c99bd95dc7bf964f755f28ab4df83fd71fcdcbd)), closes [#122](https://github.com/filipdutescu/renamer.nvim/issues/122) [#117](https://github.com/filipdutescu/renamer.nvim/issues/117) [#121](https://github.com/filipdutescu/renamer.nvim/issues/121)


### BREAKING CHANGES

* **handler:** Migrate the LSP buffer request from manual to using
`vim.lsp.buf.rename(...)` to make the request and a custom LSP response
handler to manage the custom response processing (eg setting the qf
list) and the original Neovim rename handler to apply the requested
changes.

This implies removing several existing function, doing some cleanup of
tests and updating the documentation.



## [4.0.1](https://github.com/filipdutescu/renamer.nvim/compare/v4.0.0...v4.0.1) (2022-01-15)


### Bug Fixes

* popup having `sidescrolloff` if set ([917e17d](https://github.com/filipdutescu/renamer.nvim/commit/917e17d036a68b6595e3afd209945f94607d3f9f)), closes [#101](https://github.com/filipdutescu/renamer.nvim/issues/101)
* popup not showing if renamed word appears twice ([d02d6e9](https://github.com/filipdutescu/renamer.nvim/commit/d02d6e94e968c3b0860f6b26d903147e6dd54c40)), closes [#113](https://github.com/filipdutescu/renamer.nvim/issues/113)



# [4.0.0](https://github.com/filipdutescu/renamer.nvim/compare/v3.0.3...v4.0.0) (2022-01-03)


### Bug Fixes

* **layout:** popup position when low number of lines ([0bb3b67](https://github.com/filipdutescu/renamer.nvim/commit/0bb3b67c24fa928c4e93c14b94878599e3a29518)), closes [#105](https://github.com/filipdutescu/renamer.nvim/issues/105)
* names of min and max width parameters ([d6b6aef](https://github.com/filipdutescu/renamer.nvim/commit/d6b6aefb3a3f57750671ac077f546e0d230e4974))
* **readme:** width parameter names were incorrect ([3af5113](https://github.com/filipdutescu/renamer.nvim/commit/3af51139e954b948e345974dba4f5c5360970b46))
* **rename:** responses without `changes` break `qflist` ([842b499](https://github.com/filipdutescu/renamer.nvim/commit/842b4992ecc0a566d9fcc561832d580b1d364d8c)), closes [#106](https://github.com/filipdutescu/renamer.nvim/issues/106)
* renaming leading to file rename ([814ddbb](https://github.com/filipdutescu/renamer.nvim/commit/814ddbb11602e3c8b2af166b4d1e029272ab796f)), closes [#100](https://github.com/filipdutescu/renamer.nvim/issues/100)


### Features

* add customizable min and max width ([e9d8653](https://github.com/filipdutescu/renamer.nvim/commit/e9d86536da51559974d1876ebb3b46fd096a70b3)), closes [#101](https://github.com/filipdutescu/renamer.nvim/issues/101)


### BREAKING CHANGES

* Rename `minwidth` and `maxwidth` to the naming
convention used throughout the rest of the code (snake case). The two
will now be used as `min_width` and `max_width`.

Signed-off-by: Filip Dutescu <filip.dutescu@gmail.com>



## [3.0.3](https://github.com/filipdutescu/renamer.nvim/compare/v3.0.2...v3.0.3) (2021-11-29)


### Bug Fixes

* popup showing when no LSP client attached ([1b98bae](https://github.com/filipdutescu/renamer.nvim/commit/1b98baedb0c37c196158842eb0acf70da1913f6a)), closes [#96](https://github.com/filipdutescu/renamer.nvim/issues/96)
* various errors and logs ([dd25890](https://github.com/filipdutescu/renamer.nvim/commit/dd25890dde7922c34207b9dfd7fc9f737ff1c3ee)), closes [#95](https://github.com/filipdutescu/renamer.nvim/issues/95)



## [3.0.2](https://github.com/filipdutescu/renamer.nvim/compare/v3.0.1...v3.0.2) (2021-11-23)


### Bug Fixes

* responses with `documentChanges` set qflist properly ([#90](https://github.com/filipdutescu/renamer.nvim/issues/90)) ([645e7bd](https://github.com/filipdutescu/renamer.nvim/commit/645e7bd7a18f017f05f1f5ccf54331a510c93652))



## [3.0.1](https://github.com/filipdutescu/renamer.nvim/compare/v3.0.0...v3.0.1) (2021-11-20)


### Bug Fixes

* **rename:** handle `documentChanges` LSP response ([9341d33](https://github.com/filipdutescu/renamer.nvim/commit/9341d338dd025029a904ad32784f6bd31f37d2c8)), closes [#84](https://github.com/filipdutescu/renamer.nvim/issues/84)



# [3.0.0](https://github.com/filipdutescu/renamer.nvim/compare/v2.0.0...v3.0.0) (2021-11-18)


* feat!: add custom handler for successful renames ([09030b3](https://github.com/filipdutescu/renamer.nvim/commit/09030b3286b794ccf86251604e331d5fe66ad3e8)), closes [#84](https://github.com/filipdutescu/renamer.nvim/issues/84)


### BREAKING CHANGES

* Add a new field, `handler`, representing a function
to be called after successful renames, with the LSP
`textDocument/rename` raw response.



# [2.0.0](https://github.com/filipdutescu/renamer.nvim/compare/v1.0.0...v2.0.0) (2021-11-13)


### Bug Fixes

* **qflist:** first line in file was not set correctly ([2f9e66b](https://github.com/filipdutescu/renamer.nvim/commit/2f9e66b12cc115c3c8ad3aac360fb94d76f60bce)), closes [#79](https://github.com/filipdutescu/renamer.nvim/issues/79)
* **rename:** cursor placement at word end ([37c27fa](https://github.com/filipdutescu/renamer.nvim/commit/37c27fa77571ba9a81d06e148500fad9638bbe42)), closes [#73](https://github.com/filipdutescu/renamer.nvim/issues/73)


* fix(rename)!: cursor is not placed at the end of the initial word ([a20b471](https://github.com/filipdutescu/renamer.nvim/commit/a20b471c9be59445ddd21d16885a567790dca147)), closes [#73](https://github.com/filipdutescu/renamer.nvim/issues/73)

lua/tests/health_spec.lua
local renamer = require 'renamer'
local strings = require('renamer.constants').strings

local mock = require 'luassert.mock'
local stub = require 'luassert.stub'

describe('health', function()
    describe('check', function()
        it('should generate error if a required plugin is missing', function()
            stub(vim.fn, 'health#report_error').returns()
            stub(vim.fn, 'health#report_info').returns()
            local health = require 'renamer.health'
            stub(health, '_is_plugin_installed').returns(false)
            local report_mock = mock(health.report, true)
            renamer._buffers = {}

            health.check()

            assert.spy(health._is_plugin_installed).was_called_with 'plenary'
            assert.spy(report_mock.error).was_called_with(string.format(strings.plugin_not_found_template, 'plenary'))
            assert.spy(report_mock.info).was_called_with(strings.missing_required_plugins)
            renamer._buffers = nil
            mock.revert(report_mock)
        end)

        it('should generate ok if all the required plugins are installed', function()
            local health = require 'renamer.health'
            stub(health, '_is_plugin_installed').returns(true)
            local report_mock = mock(health.report, true)
            renamer._buffers = {}

            health.check()

            assert.spy(health._is_plugin_installed).was_called_with 'plenary'
            assert.spy(report_mock.ok).was_called_with(string.format(strings.plugin_installed_template, 'plenary'))
            assert.spy(report_mock.info).was_called_with(strings.found_required_plugins)
            renamer._buffers = nil
            mock.revert(report_mock)
        end)

        it('should warn if setup was not called', function()
            local health = require 'renamer.health'
            stub(health, '_is_plugin_installed').returns(true)
            local report_mock = mock(health.report, true)
            report_mock.warn.returns()

            health.check()

            assert.spy(health._is_plugin_installed).was_called_with 'plenary'
            assert.spy(report_mock.warn).was_called_with(strings.setup_not_called)
            mock.revert(report_mock)
        end)
    end)
end)


lua/renamer/init.lua
local strings = require('renamer.constants').strings
local log = require('plenary.log').new {
    plugin = strings.plugin_name,
    level = 'warn',
}
local lsp_utils = require 'vim.lsp.util'
local popup = require 'plenary.popup'
local utils = require 'renamer.utils'
local mappings = require 'renamer.mappings'

--- @class Renamer
--- @field public title string
--- @field public padding integer[]
--- @field public min_width integer
--- @field public max_width integer
--- @field public border boolean
--- @field public border_chars string[]
--- @field public show_refs boolean
--- @field public with_qf_list boolean
--- @field public with_popup boolean
--- @field public bindings table
--- @field public handler function
--- @field private _log table
--- @field private _buffers table
--- @field private _nvim_rename_handler table
local renamer = {}
renamer._log = log

--- Setup function to be run by the user. Configures the aspect of the renamer
--- user interface. Used to change things such as the title or border of the
--- popup.
---
--- Usage:
--- <code>
--- require('renamer').setup {
---     -- The popup title, shown if `border` is true
---     title = 'Rename',
---     -- The padding around the popup content
---     padding = {
---         top = 0,
---         left = 0,
---         bottom = 0,
---         right = 0,
---     },
---     -- The minimum width of the popup
---     min_width = 15,
---     -- The maximum width of the popup
---     max_width = 45,
---     -- Whether or not to shown a border around the popup
---     border = true,
---     -- The characters which make up the border
---     border_chars = { '─', '│', '─', '│', '╭', '╮', '╯', '╰' },
---     -- Whether or not to highlight the current word references through LSP
---     show_refs = true,
---     -- Whether or not to add resulting changes to the quickfix list
---     with_qf_list = true,
---     -- Whether or not to enter the new name through the UI or Neovim's `input`
---     -- prompt
---     with_popup = true,
---     -- The keymaps available while in the `renamer` buffer. The example below
---     -- overrides the default values, but you can add others as well.
---     mappings = {
---         ['<c-i>'] = require('renamer.mappings.utils').set_cursor_to_start,
---         ['<c-a>'] = require('renamer.mappings.utils').set_cursor_to_end,
---         ['<c-e>'] = require('renamer.mappings.utils').set_cursor_to_word_end,
---         ['<c-b>'] = require('renamer.mappings.utils').set_cursor_to_word_start,
---         ['<c-c>'] = require('renamer.mappings.utils').clear_line,
---         ['<c-u>'] = require('renamer.mappings.utils').undo,
---         ['<c-r>'] = require('renamer.mappings.utils').redo,
---     },
---     -- Custom handler to be run after successfully renaming the word. Receives
---     -- the LSP 'textDocument/rename' raw response as its parameter.
---     handler = nil,
--- }
--- </code>
--- @param opts Defaults Configuration options, see `renamer.defaults`.
function renamer.setup(opts)
    opts = opts or {}

    local defaults = require 'renamer.defaults'

    renamer.title = utils.get_value_or_default(opts, 'title', defaults.title)
    renamer.padding = {
        top = utils.get_value_or_default(opts.padding, 'top', defaults.padding.top),
        left = utils.get_value_or_default(opts.padding, 'left', defaults.padding.left),
        bottom = utils.get_value_or_default(opts.padding, 'bottom', defaults.padding.bottom),
        right = utils.get_value_or_default(opts.padding, 'right', defaults.padding.right),
    }
    renamer.min_width = utils.get_value_or_default(opts, 'min_width', defaults.min_width)
    renamer.max_width = utils.get_value_or_default(opts, 'max_width', defaults.max_width)
    renamer.border = utils.get_value_or_default(opts, 'border', defaults.border)
    renamer.border_chars = utils.get_value_or_default(opts, 'border_chars', defaults.border_chars)
    renamer.show_refs = utils.get_value_or_default(opts, 'show_refs', defaults.show_refs)
    renamer.with_qf_list = utils.get_value_or_default(opts, 'with_qf_list', defaults.with_qf_list)
    renamer.with_popup = utils.get_value_or_default(opts, 'with_popup', defaults.with_popup)
    mappings.bindings = utils.get_value_or_default(opts, 'mappings', defaults.mappings)
    if opts.handler and type(opts.handler) == 'function' then
        renamer.handler = opts.handler
    end

    renamer._buffers = {}
    renamer._nvim_rename_handler = vim.lsp.handlers[strings.lsp_req_rename]
    vim.lsp.handlers[strings.lsp_req_rename] = renamer._rename_handler
    renamer._log.info(strings.finished_setup)
end

--- Function that renames the word under the cursor, using Neovim's built in
--- LSP features (`vim.lsp.buf_request()`). Creates a popup next to the cursor,
--- starting at the beginning of the word.
---
--- The popup is drawn below the current line, if there is enough space,
--- otherwise on the one above it.
---
--- Usage:
--- <code>
--- require('renamer').rename()
--- </code>
---
--- To rename without having the existing name in the popup, use the following:
---
--- <code>
--- require('renamer').rename {
---     empty = true,
--- }
--- </code>
--- @param opts table Rename specific options (ie: `empty`).
--- @return integer prompt_window_id
--- @return table prompt_window_opts @Keys: opts, border_opts
function renamer.rename(opts)
    opts = opts or { empty = false }

    local lsp_clients = vim.lsp.buf_get_clients(0)
    if lsp_clients == nil or #lsp_clients < 1 then
        renamer._log.error(strings.no_lsp_client_found_err)
    end

    local cword = vim.fn.expand(strings.cword_keyword)
    local win_height = vim.api.nvim_win_get_height(0)
    local win_width = vim.api.nvim_win_get_width(0)
    local popup_opts = renamer._create_default_popup_opts(cword)
    local padding_top_bottom = renamer.padding.top + renamer.padding.bottom
    local padding_left_right = renamer.padding.left + renamer.padding.right
    local is_height_too_short = renamer.border == true and win_height < 4 + padding_top_bottom
        or not (renamer.border == true) and win_height < 2 + padding_top_bottom
    local is_width_too_short = renamer.border == true and win_width < popup_opts.minwidth + 2 + padding_left_right
        or not (renamer.border == true) and win_width < popup_opts.minwidth + padding_left_right
    local line, col = renamer._get_cursor()
    local word_start, _ = utils.get_word_boundaries_in_line(vim.api.nvim_get_current_line(), cword, col + 1)
    if word_start == nil then
        renamer._log.info(strings.invalid_cursor_position_err)
        return
    end
    local prompt_col_no, prompt_line_no = col - word_start + 1, 2
    local line_count = vim.api.nvim_buf_line_count(0)
    local lines_from_win_end = line_count - line
    local width = 0

    if renamer.title then
        width = #renamer.title + 4
    end
    if width >= win_width or #cword > width then
        width = #cword + 1
    end
    if not (renamer.border == true) then
        prompt_line_no = 1
    end
    if lines_from_win_end < prompt_line_no + 1 and line_count >= win_height - 1 then
        prompt_line_no = -prompt_line_no
    end
    if word_start + width >= win_width then
        prompt_col_no = prompt_col_no + width - win_width + word_start
        if renamer.border == true then
            prompt_col_no = prompt_col_no + 4
        end
    end

    renamer._document_highlight()

    popup_opts.width = width
    popup_opts.line = (prompt_line_no >= 0 and strings.plenary_popup_cursor_plus or strings.plenary_popup_cursor)
        .. prompt_line_no
    popup_opts.col = strings.plenary_popup_cursor_minus .. prompt_col_no
    popup_opts.initial_pos = {
        word_start = word_start,
        col = col,
        line = line,
    }

    if is_height_too_short or is_width_too_short or renamer.with_popup == false then
        if is_height_too_short or is_width_too_short then
            renamer._log.error(strings.not_enough_space_err)
        end

        renamer._input_lsp_rename(cword, popup_opts.initial_pos)
        renamer._clear_references()
        return
    end

    local popup_content = cword
    if opts.empty == true then
        popup_content = ''
    end
    local prompt_win_id, prompt_opts = popup.create(popup_content, popup_opts)

    renamer._buffers[prompt_win_id] = {
        opts = popup_opts,
        border_opts = prompt_opts.border,
    }
    renamer._log.fmt_info(strings.created_popup_template, vim.inspect(renamer._buffers[prompt_win_id]))

    renamer._setup_window(prompt_win_id)
    renamer._set_cursor_to_popup_end()
    renamer._log.trace(strings.finished_popup_setup)
    renamer._set_prompt_win_style(prompt_win_id)
    renamer._log.trace(strings.finished_popup_styling)

    return prompt_win_id, renamer._buffers[prompt_win_id]
end

function renamer.on_submit(window_id)
    if window_id and renamer._buffers and renamer._buffers[window_id] then
        local opts = renamer._buffers[window_id].opts
        local pos = opts and opts.initial_pos
        local buf_id = vim.api.nvim_win_get_buf(window_id)
        local new_word = vim.api.nvim_buf_get_lines(buf_id, -2, -1, false)[1]

        if new_word == nil then
            new_word = ''
        end

        renamer._delete_autocmds()
        if not (new_word == '') then
            new_word = string.gsub(new_word, '^%s*(.-)%s*$', '%1')
            renamer._log.fmt_info(strings.submitted_word_template, new_word)
            renamer.on_close(window_id)
            renamer._lsp_rename(new_word, pos)
        else
            renamer.on_close(window_id)
            renamer._log.fmt_error(strings.rename_to_empty_string_err_template, opts.initial_word)
        end
    end
end

function renamer.on_close(window_id)
    local settings = renamer._buffers and renamer._buffers[window_id]
    local border_win_id = settings and settings.border_opts and settings.border_opts.win_id
    local initial_mode = settings and settings.opts and settings.opts.initial_mode
    local pos = settings and settings.opts and settings.opts.initial_pos

    renamer._delete_window(window_id)
    renamer._log.fmt_info(strings.deleted_window_template, window_id)
    renamer._delete_window(border_win_id)
    renamer._log.fmt_info(strings.deleted_window_border_template, border_win_id)

    renamer._clear_references()
    if initial_mode and not string.match(initial_mode, strings.insert_mode_short_string) then
        vim.api.nvim_command(strings.stopinsert_command)
    end

    if pos then
        local col = pos.col
        if initial_mode and not string.match(initial_mode, strings.insert_mode_short_string) then
            col = col + 1
        end
        vim.api.nvim_win_set_cursor(0, { pos.line, col })
    end
end

function renamer._create_default_popup_opts(cword)
    local p = renamer.padding
    return {
        title = renamer.title,
        titlehighlight = strings.highlight_title,
        padding = { p.top, p.right, p.bottom, p.left },
        border = renamer.border,
        borderchars = renamer.border_chars,
        highlight = strings.highlight_normal,
        borderhighlight = strings.highlight_border,
        minwidth = renamer.min_width,
        maxwidth = renamer.max_width,
        minheight = 1,
        posinvert = false,
        cursor_line = true,
        enter = true,
        initial_word = cword,
        initial_mode = vim.api.nvim_get_mode().mode,
    }
end

function renamer._get_cursor()
    -- [[
    -- `cursor` is an array of two elements with the following semnification:
    --      - *`cursor[1]`* - the current cursor line
    --      - *`cursor[2]`* - the current cursor column
    -- ]]
    local cursor = vim.api.nvim_win_get_cursor(0)

    return cursor[1], cursor[2]
end

function renamer._set_cursor_to_popup_end()
    local popup_win_content = vim.api.nvim_buf_get_lines(0, -2, -1, true)
    local cursor_line, cursor_col = #popup_win_content, #popup_win_content[#popup_win_content]

    cursor_line = cursor_line + renamer.padding.top
    cursor_col = cursor_col - renamer.padding.right

    vim.api.nvim_win_set_cursor(0, { cursor_line, cursor_col })
end

function renamer._setup_window(prompt_win_id)
    local prompt_buf_id = vim.api.nvim_win_get_buf(prompt_win_id)

    vim.api.nvim_win_set_option(prompt_win_id, strings.win_opt_wrap, false)
    vim.api.nvim_win_set_option(prompt_win_id, strings.win_opt_winblend, 0)
    vim.api.nvim_win_set_option(prompt_win_id, strings.win_opt_sidescrolloff, 0)

    vim.api.nvim_command(strings.startinsert_command)
    renamer._create_autocmds(prompt_win_id)

    vim.api.nvim_buf_set_keymap(
        prompt_buf_id,
        strings.insert_mode_short_string,
        strings.cr_keyword,
        string.format(strings.submit_key_press_command, prompt_win_id),
        { noremap = true }
    )
    vim.api.nvim_buf_set_keymap(
        prompt_buf_id,
        strings.insert_mode_short_string,
        strings.esc_keyword,
        string.format(strings.cancel_key_press_command, prompt_win_id),
        { noremap = true }
    )
    mappings.register_bindings(prompt_buf_id)
end

function renamer._set_prompt_win_style(prompt_win_id)
    if prompt_win_id then
        if renamer._buffers and renamer._buffers[prompt_win_id] then
            local opts = renamer._buffers[prompt_win_id]
            local border_win_id = opts.border_opts and opts.border_opts.win_id

            renamer._set_prompt_border_win_style(border_win_id)
        end
    end
end

function renamer._set_prompt_border_win_style(prompt_border_win_id)
    if prompt_border_win_id then
        vim.api.nvim_win_set_option(prompt_border_win_id, strings.win_opt_wrap, false)
        vim.api.nvim_win_set_option(prompt_border_win_id, strings.win_opt_winblend, 0)
    end
end

function renamer._create_autocmds(prompt_win_id)
    local on_leave = string.format(strings.autocmd_buf_leave_template, prompt_win_id)

    vim.cmd(strings.augroup_start)
    vim.cmd(strings.augroup_reset)
    vim.cmd(on_leave)
    vim.cmd(strings.augroup_end)
end

function renamer._delete_autocmds()
    vim.cmd(strings.augroup_start)
    vim.cmd(strings.augroup_reset)
    vim.cmd(strings.augroup_end)
end

function renamer._input_lsp_rename(cword, position)
    local new_word = vim.fn.input(string.format(strings.input_prompt_template, cword))

    if new_word and not (new_word == '') then
        renamer._lsp_rename(new_word, position)
    end
end

function renamer._lsp_rename(word, pos)
    renamer._current_op = {
        pos = pos,
        word = word,
    }
    renamer._lsp_rename_internal(word)
end

-- Since there is no way to mock `vim.lsp.buf.rename`, this function is used as a replacement.
function renamer._lsp_rename_internal(word)
    vim.lsp.buf.rename(word)
end

function renamer._rename_handler(...)
    local resp
    local method
    local err = select(1, ...)
    local is_new = not select(4, ...) or type(select(4, ...)) ~= 'number'

    if is_new then
        method = select(3, ...).method
        resp = select(2, ...)
    else
        method = select(2, ...)
        resp = select(3, ...)
    end

    if err then
        renamer._log.error(err)
        return
    end
    if resp == nil then
        renamer._log.warn(strings.nil_lsp_response_err)
        return
    end

    if method == strings.lsp_req_rename then
        renamer._nvim_rename_handler(...)
    else
        vim.lsp.handlers[method](...)
    end

    if renamer.with_qf_list then
        local changes = resp.changes or {}
        if resp.documentChanges then
            for _, change in ipairs(resp.documentChanges) do
                if change.textDocument and change.textDocument.uri then
                    changes[change.textDocument.uri] = change.edits
                end
            end
            for _, change in ipairs(resp.documentChanges) do
                if change.newUri and change.oldUri and changes[change.oldUri] then
                    local edits = changes[change.oldUri]
                    changes[change.newUri] = edits
                    changes[change.oldUri] = nil
                end
            end
        end

        utils.set_qf_list(changes)
    end
    if renamer._current_op and renamer._current_op.pos then
        local col = renamer._current_op.pos.word_start + #renamer._current_op.word - 1
        local mode = vim.api.nvim_get_mode().mode
        if mode and not string.match(mode, strings.insert_mode_short_string) then
            col = col - 1
        end
        vim.api.nvim_win_set_cursor(0, { renamer._current_op.pos.line, col })
        renamer._current_op = {}
    end
    if renamer.handler then
        renamer.handler(resp)
    end
end

function renamer._document_highlight()
    if renamer.show_refs then
        renamer._document_highlight_internal()
    end
end

-- Since there is no way to mock `vim.lsp.buf.document_highlight`, this function is used as a replacement.
function renamer._document_highlight_internal()
    vim.lsp.buf.document_highlight()
end

function renamer._delete_window(win_id)
    if win_id and vim.api.nvim_win_is_valid(win_id) then
        local buf_id = vim.api.nvim_win_get_buf(win_id)
        if
            buf_id
            and vim.api.nvim_buf_is_valid(buf_id)
            and not vim.api.nvim_buf_get_option(buf_id, strings.buf_opt_buflisted)
        then
            vim.api.nvim_command(string.format(strings.buf_delete_command_template, buf_id))
        end

        if vim.api.nvim_win_is_valid(win_id) then
            vim.api.nvim_win_close(win_id, true)
            if not pcall(vim.api.nvim_win_close, win_id, true) then
                renamer._log.trace(string.format(strings.failed_to_close_window_err_template, win_id))
            end
        end
    end

    if win_id and renamer._buffers and renamer._buffers[win_id] then
        renamer._buffers[win_id] = nil
        renamer._delete_autocmds()
    end
end

function renamer._clear_references()
    if renamer.show_refs then
        renamer._clear_references_internal()
    end
end

-- Since there is no way to mock `vim.lsp.buf.clear_references`, this function is used as a replacement.
function renamer._clear_references_internal()
    vim.lsp.buf.clear_references()
end

function renamer._apply_workspace_edit(resp)
    lsp_utils.apply_workspace_edit(resp)
end

return renamer
